msprime/mutations.py

```diff
@@ -1183,26 +1184,29 @@ def sim_mutations(
     then the same mutations will be generated. If no random seed is specified
     then one is generated automatically.
 
-    By default, sites and mutations in the input tree sequence are
-    discarded. If the ``keep`` parameter is true, however, *additional*
-    mutations are simulated. Under the infinite sites mutation model, all new
-    mutations generated will occur at distinct positions from each other and
-    from any existing mutations (by rejection sampling). Furthermore, if sites
-    are discrete, trying to simulate mutations at time periods that are older
-    than mutations kept from the original tree sequence is an error, because
-    this would create an extra transition (from the new allele to the old
-    one below it) that may be incorrect according to the model of mutation.
-    Under a state-independent mutation model, however (e.g., Jukes-Cantor),
-    there is no problem, and ``kept_mutations_before_end_time=True`` may be
-    set to allow adding new mutations around or above existing ones.
-
     The time interval over which mutations can occur may be controlled
     using the ``start_time`` and ``end_time`` parameters. The ``start_time``
     defines the lower bound (in time-ago) on this interval and ``max_time``
     the upper bound. Note that we may have mutations associated with
     nodes with time <= ``start_time`` since mutations store the node at the
     bottom (i.e., towards the leaves) of the branch that they occur on.
 
+    If the tree sequence already has mutations, these are by default retained,
+    but can be discarded by passing ``keep=False``. However, adding new
+    mutations to a tree sequence with existing mutations must be done with
+    caution, since it can lead to incorrect or nonsensical results if mutation
+    probabilities differ by ancestral state. (As an extreme example, suppose
+    that X->Y and X->Z are allowable transitions, but Y->Z is not. If a branch
+    already has an X->Y mutation on it, then calling `sim_mutations(...,
+    keep=True)` might insert an X->Z mutation above the existing mutation, thus
+    implying the impossible chain X->Y->Z.)  For this reason, if this method
+    attempts to add a new mutation ancestral to any existing mutation, an error
+    will occur, unless ``add_ancestral=True``. The ``add_ancestral`` parameter
+    has no effect if ``keep=False``. In summary, to add more mutations to a
```

Comment:
Edits:
```
Furthermore, even if ``add_ancestral`` is True, this will throw an error if there are any mutations that result in a silent transition (e.g., placing a mutation to A above an existing mutation to A), since silent transitions are not allowed in tskit.
[In summary ... or set ``add_ancestral=True``]
and ensure that silent mutations cannot occur.
```

Reply:
Done, but I don't know how a user could ensure that silent mutations cannot occur (since we require the same alphabet)

Reply:
I've changed the last bit to "and be prepared for ocassional errors when silent mutations arise".

I think would lead to a lot of confusion, but let's get this merged and follow up.

Reply:
>  I don't know how a user could ensure that silent mutations cannot occur (since we require the same alphabet)

It'd have to be an infinite alleles model (that doesn't look at the parent allele).